WreckEngine/Text.cpp

```diff
@@ -1,198 +1,549 @@
+    FT_Glyph glyph = nullptr;
+    if (!FT_Load_Char(fontFace, ' ', FT_LOAD_DEFAULT) &&
+        !FT_Get_Glyph(fontFace->glyph, &glyph))
+    {
+        spaceWidth = fontFace->glyph->metrics.horiAdvance * kerningScale;
+    }
+
+    // Get the line height
+    lineHeight = fontFace->size->metrics.height * kerningScale;
+
+    // TODO - We may not want to immediately load glyphs here, or we may want to specify an overload for a glyph range
+    loadGlyphs();
+}
+
+void Text::FontFace::loadGlyphs()
+{
+    // Loads all basic, printable ASCII characters
+    loadGlyphRange(0, 255);
+}
+
+void Text::FontFace::loadGlyphRange(uint32_t begin, uint32_t end)
+{
+    // TODO - (M)SDF conversion?
+    // TODO - Currently assumes the texture is empty. Expected functionality would be to have
+    //        multiple glyph ranges loaded at a time
+
+    constexpr int rectPaddingPerSide = 1;
+    constexpr int rectPadding = rectPaddingPerSide * 2;
+
+    // Load each of the glyphs in the range
+    std::vector<stbrp_rect> packRects;
+    std::vector<GlyphData> loadedGlyphs;
+    for (uint32_t cp = begin; cp <= end; ++cp)
+    {
+        // Load code point
+        if (FT_Load_Char(fontFace, cp, FT_LOAD_RENDER | FT_LOAD_FORCE_AUTOHINT))
+        {
+            std::cout << "Could not load code point " << cp << " (" << static_cast<char>(cp) << ')' << std::endl;
+            continue;
+        }
+
+        auto& glyph = fontFace->glyph;
+        auto& bitmap = glyph->bitmap;
+
+        const auto bitmapWidth = bitmap.width;
+        const auto bitmapHeight = bitmap.rows;
+
+        // Create pack rect (inflate it a bit for padding)
+        stbrp_rect rect = {0};
+        rect.id = cp;
+        rect.w = static_cast<stbrp_coord>(bitmapWidth + rectPadding);
+        rect.h = static_cast<stbrp_coord>(bitmapHeight + rectPadding);
+        packRects.push_back(rect);
+
+        // Create glyph data
+        GlyphData glyphData;
+        auto& glyphMetrics = fontFace->glyph->metrics;
+        glyphData.cp = cp;
+        glyphData.bitmap.resize(rect.w * rect.h, 0); // texture is only alpha values
+        glyphData.glyph.advance = glyphMetrics.horiAdvance * kerningScale;
+        glyphData.glyph.bearing = {
+            glyphMetrics.horiBearingX * kerningScale,
+            glyphMetrics.horiBearingY * kerningScale
+        };
+        glyphData.glyph.size = {
+            glyphMetrics.width        * kerningScale,
+            glyphMetrics.height       * kerningScale
+        };
+
+        // Copy glyph buffer data (all pixels are white with varying alpha levels)
+        const unsigned char* pixels = bitmap.buffer;
+        auto& glyphBuffer = glyphData.bitmap;
+        for (uint32_t y = 0; y < bitmapHeight; ++y)
+        {
+            for (uint32_t x = 0; x < bitmapWidth; ++x)
+            {
+                const size_t index = (x + rectPaddingPerSide) + (y + rectPaddingPerSide) * rect.w;
+                if (bitmap.pixel_mode == FT_PIXEL_MODE_MONO)
```

Comment:
this could be moved outside the loop for better performance, albeit uglier looking code